### PR TITLE
Support hosts list configs via env variables

### PR DESCRIPTION
**Pull request by Tedderouni**
This is my attempt to support passing in a hosts list array via an environment variable, formatted as a comma-separated list.  There have been multiple threads about this (primarily #8157, also #6366, #6665, and maybe more) with various implementation preferences, but the ones I've seen have all petered out, so I'm hoping to get this going.  There are a few potential discussion points here, so feedback is welcome.

First, this will break any existing configs that already pass in a host value that contains a comma via an env variable.  IMHO this is probably rare.  Commas are technically valid in a subpath, but [according to the docs](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-hosts) they should be encoded (along with any other special characters).  However raw commas do work anyway before this change.  And actually, encoded commas don't work - there seems to be a bug where URLs are not actually decoded.  Because of this, I'm unescaping in my code (in order to allow the encoded commas in the subpath), but if this gets fixed in the rest of the code, it shouldn't be needed here.  I'll open a separate issue for this.

I decided to support only the hosts parameter, rather than allow any environment variable to be split this way - my thought was to limit breaking impact because other config parameters might have desirable commas, which would then need to be escaped, so this could break more existing configs.  If we want to add more parameters to be parsed like this, we can easily add their names/validators to the conditional.

Per @jordansissel's comments in #8157, my implementation allows the host list to be specified as an environment variable (and also in the secret store), but doesn't allow a literal list in the Logstash config.  I wasn't able to use only the validator because the first call to deep_replace happens before validation, so it's not aware of the validator yet, so my change receives either the parameter name or the validator and checks both values, looking for either "hosts" or :uri, and then splits the value on commas.

If this approach gets approved, I'll also update the docs to mention support for this syntax.

### CODE REVIEW

**Comment by Tedderouni** [comment]
@jsvd Has anyone had a chance to look over this?  I'm very interested in being able to use this in an official build.  Thanks!

File: logstash-core/lib/logstash/util/substitution_variables.rb [hunk 1]
```diff
@@ -50,6 +50,20 @@ def replace_placeholders(value)
         raise LogStash::ConfigurationError, "Cannot evaluate `#{placeholder}`. Replacement variable `#{name}` is not defined in a Logstash secret store " +
             "or as an Environment entry and there is no default value given."
       end
+
+      if [ "hosts", :uri ].include?(name_or_validator.join)
```
Comment (fbaligand) [inline logstash-core/lib/logstash/util/substitution_variables.rb]: Here, it would be great to add also `:array`

Comment (Tedderouni) [reply logstash-core/lib/logstash/util/substitution_variables.rb]: Thanks @fbaligand !  I've added it, let me know if this isn't what you meant.

Comment (fbaligand) [reply logstash-core/lib/logstash/util/substitution_variables.rb]: This is exactly what I meant :)
=> Nice!

**Comment by Tedderouni** [comment]
@fbaligand Thanks! Do you see anything else needed here before this can be merged?

**Comment by fbaligand** [comment]
Hi,
For me, that's OK.
Unfortunately, I'm just a Logstash contributor (as you). I haven't rights to merge this PR.
Sorry.

**Comment by Tedderouni** [comment]
Yep, no worries.  I appreciate the feedback!

**Comment by fbaligand** [comment]
BTW, I would love that this PR is merged!
It would perfectly complete my initial work some years ago:
https://github.com/elastic/logstash/pull/4710

**Comment by tobyDickinson** [comment]
Incredibly keen for this PR also

**Comment by colinsurprenant** [comment]
👀

**Comment by colinsurprenant** [comment]
First, thanks @Tedderouni for your contribution and seconds, sorry for the very long delay in following up. 

While investigating this, I see that we have merged #12051 and was wondering if that actually solved this use-case?  Let me know if that PR provides the functionality your are looking for or not.

**Comment by fbaligand** [comment]
Hi @colinsurprenant,

The PR you mention is a great enhancement.
But it does not solve all list configs via env variables, as explained in these 2 comments:
- https://github.com/elastic/logstash/issues/6366#issuecomment-650520649
- https://github.com/elastic/logstash/issues/6366#issuecomment-662732939

**Comment by colinsurprenant** [comment]
After reviewing the history and challenges about this here are my thoughts and suggestions.

This PR tackles a single use-case, the hosts list option to try and minimise breaking changes. While I agree this solves a very common issue, it only looks at the hosts list issue. Ideally I would prefer we agree on a more holistic solution that aims to solve the problem for all string expansions into lists. 

I do not think it is possible to do that into a backward compatible way. 

We could look at that with two horizons: 
- Establish the global solution which will likely be a breaking change for **v8.0.0**, that will include string expansion of any list-type options  for both regular config files and plugin options in pipeline definitions.
- Start implementing in **v7** and make the behaviour possibly available as an opt-in in v7.

**Comment by fbaligand** [comment]
Great!
I fully agree with you @colinsurprenant, that the interesting point is to address every list option, and not only hosts list. 

And I find very nice your idea to have an opt-in flag in v7 and a breaking change in v8!

**Comment by Tedderouni** [comment]
Thank you for the replies.  I finally had a chance to test this use case on v7.9 using space-delimited hostnames, and it worked, so it seems #12051 solved this.  Closing this PR as the goal has been met.

I also agree @colinsurprenant that a broader solution would be ideal.  Looking forward to seeing it!

**Pull request closed without merging**